draft-ietf-tls-esni.md

```diff
@@ -724,21 +741,26 @@ MAY offer to resume sessions established without ECH.
 
 # Server Behavior {#server-behavior}
 
-Servers that that implement the ECH extension play one of two roles, depending
-on the form of the ECH extension in the ClientHello. If the extension value is
-non-empty, the server acts as a client-facing server and proceeds as described
-in {{client-facing-server}} to extract a ClientHelloInner if available. If the
-extension value is empty, the server acts as a backend server and proceeds as
+Servers that that support ECH play one of two roles, depending on the extension
+in the ClientHello. If the "encrypted_client_hello" extension is present, the
+server acts as a client-facing server and proceeds as described in
+{{client-facing-server}} to extract a ClientHelloInner, if available. If the
+"is_client_hello_inner" extension is present and the "encrypted_client_hello"
+extension is not present, the server acts as a backend server and proceeds as
```

Comment:
What if both extensions are present? I suppose the server MUST abort with illegal_parameter in that case? (I see below that it's suggested for servers to ignore the inner one, but it seems better to fail loudly here, since this is not permitted client behavior.)

Reply:
> ... but it seems better to fail loudly here, since this is not permitted client behavior.

This seems to be true in general, but elsewhere in the draft we deal with client misbehavior by rejecting. For example, if ClientECH.cipher_suite isn't a valid cipher suite for the ECHConfig indicated by ClientECH.config_id, then the client-facing server skips this ECHConfig and moves on to the next. This done for "don't stick out" reasons. I think it makes sense to follow the same pattern here.

TL;DR: I think the text is better as-is.

Reply:
But a difference in algorithm is not the same as an actual implementation bug on the client side. It could be because of a server misconfiguration. This edge case (sending both extensions) is a flat out bug, no?

Reply:
It doesn't add anything to the protocol, but should it be fatal?

Reply:
I would argue that choosing a cipher suite other than one that is indicated by the ECHConfig is also an implementation bug.

Reply:
I think other people need to weigh in here. @davidben, penny for your thoughts?

Reply:
I think let's require failure.